packages/client/src/runtime/core/types/Result.ts

```diff
@@ -82,31 +81,25 @@ type SelectablePayloadFields<K extends PropertyKey, O> =
   | { composites: { [k in K]: O } }
 
 // prettier-ignore
-type SelectField<P extends SelectablePayloadFields<any, any>, K extends PropertyKey> = 
+type SelectField<P extends SelectablePayloadFields<any, any>, K extends PropertyKey> =
   P extends { objects: Record<K, any> } 
   ? P['objects'][K]
   : P extends { composites: Record<K, any> }
     ? P['composites'][K]
     : never
 
 // prettier-ignore
-export type DefaultSelection<P> = P extends Payload 
-  ? P['scalars'] & UnwrapPayload<P['composites']>
-  : P
+export type DefaultSelection<P> = UnwrapPayload<{ default: P }>['default']
 
 // prettier-ignore
-type UnwrapPayload<P> = {
-  [K in keyof P]:
-    P[K] extends Payload[]
-    ? UnwrapPayload<P[K]>
-    : P[K] extends Payload 
-      ? P[K]['scalars'] & UnwrapPayload<P[K]['composites']> 
-      : P[K] extends infer O | null 
-        ? O extends Payload 
-          ? (O['scalars'] & UnwrapPayload<O['composites']>) | null
-        : P[K]
-      : P[K]
-} & unknown
+type UnwrapPayload<P> = {} extends P ? unknown : {
+  [K in keyof P]: 
+    P[K] extends { scalars: infer S, composites: infer C }[]
+    ? Array<S & UnwrapPayload<C>>
+    : P[K] extends { scalars: infer S, composites: infer C } | null
+      ? S & UnwrapPayload<C> | Select<P[K], null>
+      : never
```

Comment:
Is case for non nullable `P[K]` missing here?

Reply:
I think it would be covered since unions are distributed, wdyt? The simplest example I have in mind something like:

```ts
type Test<O> = {
  [K in keyof O]: O[K] extends number | null ? true : false
}

type result = Test<{ a: null, b: number, c: number | null }>
```

Reply:
Yes, better to test. I'll check and add as needed.

Reply:
We have a test for a nested required composite [here](https://github.com/prisma/prisma/blob/9bab0b986d7e021c86069a42c1a950788891bcf9/packages/client/tests/functional/composites/selection/tests.ts/#L101)

Reply:
I forgot to mention in the PR, I wanted to point out that sometimes when recursive types shows up this way it can be a sign that the compiler is too eagerly evaluating and can (not always, but is a good proxy when developing) be related to other performance issues/responsiveness down the line. Also, when you hover over the type, that experience should be reflected there as well.

This is the previous implementation as you and I implemented it
![image](https://github.com/prisma/prisma/assets/18401805/fe1a8fa8-5fc3-4c92-853b-30d909b877b5)

This is how the refactored/proposed replacement is evaluated (well, not over-evaluated)
![image](https://github.com/prisma/prisma/assets/18401805/d0d13c17-17c9-45c3-91a3-05205cfc15c7)

Reply:
(I have started working on type performance tests btw, but work is not yet ready)